### PR TITLE
[SPARK-39383][SQL] Support DEFAULT columns in ALTER TABLE ALTER COLUMNS to V2 data sources

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Extend DEFAULT column support in ALTER TABLE ALTER COLUMNS commands to include V2 data sources.

(Note: this depends on https://github.com/apache/spark/pull/36771.)

Example:

```
> create or replace table t (a string default 'abc', b string) using $v2Source
> insert into t values (default)
> alter table t alter column b set default 'def'
> insert into t values ("ghi")
> Select * from t
"abc", null,
"ghi", "def"
```

### Why are the changes needed?

This makes V2 data sources easier to use and extend.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

This PR includes new test coverage.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala [hunk 1]
```diff
@@ -519,6 +519,7 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
   private[sql] lazy val existenceDefaultValues: Array[Any] = getExistenceDefaultValues(this)
   private[sql] lazy val existenceDefaultsBitmask: Array[Boolean] = getExistenceDefaultsBitmask(this)
   private[sql] lazy val hasExistenceDefaultValues = existenceDefaultValues.exists(_ != null)
+  private[sql] lazy val defaultColumnAnalyzer: Analyzer = DefaultColumnAnalyzer
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala:522]: Shall we remove the changes in this file?

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala]: Sure, thanks for catching this, done.

**Comment by dtenedor** [comment]
@gengliangwang I have merged in the changes from the base branch, this PR is much simpler now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 2]
```diff
@@ -203,6 +203,14 @@ private[sql] object CatalogV2Util {
               })
           }
 
+        case update: UpdateColumnDefaultValue =>
+          replace(schema, update.fieldNames, field =>
+            if (update.newDefaultValue().nonEmpty) {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:208]: What if the new default value is an empty string? Do we just clear the default value?

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:208]: Oh wait, it seems that will be "\"\""
Then when will the default value get cleared?

**Comment by gengliangwang** [comment]
LGTM except for one comment

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:208]: Good question. The new DEFAULT value string will be non-empty for any DDL commands that set the default value, such as "ALTER TABLE t ALTER COLUMN c SET DEFAULT ..." (this is enforced by the parser). On the other hand, commands that drop the default value such as "ALTER TABLE t ALTER COLUMN c DROP DEFAULT" will set this string to empty. I added this in a comment here to help explain.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 3]
```diff
@@ -203,6 +203,18 @@ private[sql] object CatalogV2Util {
               })
           }
 
+        case update: UpdateColumnDefaultValue =>
+          replace(schema, update.fieldNames, field =>
+            // The new DEFAULT value string will be non-empty for any DDL commands that set the
+            // default value, such as "ALTER TABLE t ALTER COLUMN c SET DEFAULT ..." (this is
+            // enforced by the parser). On the other hand, commands that drop the default value such
+            // as "ALTER TABLE t ALTER COLUMN c DROP DEFAULT" will set this string to empty.
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:211]: How about testing "DROP DEFAULT" as well?  I will merge this one after the new case added

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:211]: Sure, good idea. This is done.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master

**Pull request closed without merging**